Build fails

https://github.com/equinor/Carolina/actions/runs/19532245635/job/55917740875

This is very likely due to the fact that the image for `macos-14` contains `cmake==4.1.2.`.
https://github.com/actions/runner-images/blob/macos-14-arm64/20251111.0092/images/macos/macos-14-arm64-Readme.md

The `ubuntu-latest` image contains `cmake==3.31.6` and works fine.

Suggest to either provide another cmake version to the runner, or adjust `CMAKE_MINIMUM_REQUIRED`.

```
CMake Error at /tmp/build/dakota-6.21.0-public-src-cli/packages/external/trilinos/cmake/tribits/core/config_tests/fmangle/CMakeLists.txt:40 (CMAKE_MINIMUM_REQUIRED):
  Compatibility with CMake < 3.5 has been removed from CMake.
```

We should also use `macos-latest` instead of `macos-14` runners specifically.

--- a/.github/workflows/bundle_with_dakota_caller.yml
+++ b/.github/workflows/bundle_with_dakota_caller.yml
@@ -36,7 +36,7 @@ jobs:
       NEEDS_REBUILD: ${{ env.NEEDS_REBUILD }}
 
     steps:
-      - uses: actions/cache@v4
+      - uses: actions/cache@v5
         name: Check cache for already built wheels
         id: cache-package
         with:
@@ -76,7 +76,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: "Download built artifacts"
-        uses: actions/download-artifact@v4
+        uses: actions/download-artifact@v8
         with:
           path: artifacts
     
@@ -88,12 +88,12 @@ jobs:
       
       - name: Cache wheels folder
         if: steps.cache-package.outputs.cache-hit != 'true'
-        uses: actions/cache/save@v4
+        uses: actions/cache/save@v5
         with:
           path: dist
           key: carolina_wheels_boost${{ needs.build-settings.outputs.BOOST_VERSION }}_dakota${{ needs.build-settings.outputs.DAKOTA_VERSION }}
 
-      - uses: actions/cache/restore@v4
+      - uses: actions/cache/restore@v5
         id: restore-cached-package
         with:
           key: carolina_wheels_boost${{ needs.build-settings.outputs.BOOST_VERSION }}_dakota${{ needs.build-settings.outputs.DAKOTA_VERSION }}
@@ -102,8 +102,9 @@ jobs:
       - name: Validate dist with twine check
         run: |
           ls dist
+          pip install -U pip
           pip install twine
-          pip install packaging==25.0
+          pip install -U packaging
           twine check dist/*
 
           # remove carolina dev-wheels if any
@@ -115,7 +116,7 @@ jobs:
           ls dist
 
       - name: Publish to pypi
-        uses: pypa/gh-action-pypi-publish@v1.12.4
+        uses: pypa/gh-action-pypi-publish@v1.13.0
         if: >
           (github.event_name == 'workflow_dispatch' && inputs.publish) ||
           (github.event_name == 'push' && startsWith(github.ref, 'refs/tags'))

--- a/.github/workflows/bundle_with_dakota_linux.yml
+++ b/.github/workflows/bundle_with_dakota_linux.yml
@@ -30,7 +30,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: [ '3.11', '3.12', '3.13' ]
+        python-version: [ '3.11', '3.12', '3.13', '3.14' ]
         os: [ ubuntu-24.04 ]
     runs-on: ${{ matrix.os }}
     env:
@@ -38,11 +38,11 @@ jobs:
       NO_PROJECT_RES: ${{ inputs.NO_PROJECT_RES }}
 
     steps:
-    - uses: actions/checkout@v4
+    - uses: actions/checkout@v6
       with:
         fetch-depth: 0
 
-    - uses: actions/cache@v4
+    - uses: actions/cache@v5
       id: cache-package-check
       with:
         key: carolina_deps_${{ matrix.os }}_python-${{ matrix.python-version }}_boost-${{ inputs.BOOST_VERSION }}_dakota-${{ inputs.DAKOTA_VERSION }}
@@ -50,7 +50,7 @@ jobs:
 
     - name: Build Boost & Dakota
       if: steps.cache-package-check.outputs.cache-hit != 'true'
-      uses: docker://quay.io/pypa/manylinux2014_x86_64
+      uses: docker://quay.io/pypa/manylinux_2_28
       env:
         PYTHON_VERSION: ${{ matrix.python-version }}
         BOOST_VERSION: ${{ inputs.BOOST_VERSION }}
@@ -59,7 +59,7 @@ jobs:
         entrypoint: /bin/bash
         args: '-c "sh dakota_manylinux_install_files/build_deps_gha.sh ${{ matrix.python-version }}"'
 
-    - uses: actions/cache/save@v4
+    - uses: actions/cache/save@v5
       if: steps.cache-package-check.outputs.cache-hit != 'true'
       id: cache-package-store
       with:
@@ -67,7 +67,7 @@ jobs:
         path: ./deps_build
 
     - name: Build Linux Wheel
-      uses: docker://quay.io/pypa/manylinux2014_x86_64
+      uses: docker://quay.io/pypa/manylinux_2_28
       env:
         PYTHON_VERSION: ${{ matrix.python-version }}
         BOOST_VERSION: ${{ inputs.BOOST_VERSION }}
@@ -77,13 +77,13 @@ jobs:
         args: '-c "sh dakota_manylinux_install_files/build_wheels_gha.sh ${{ matrix.python-version }}"'
 
     - name: Checkout Carolina
-      uses: actions/checkout@v4
+      uses: actions/checkout@v6
       with:
         fetch-depth: 0
         path: carolina-main
 
     - name: Set up Python ${{ matrix.python-version }}
-      uses: actions/setup-python@v5
+      uses: actions/setup-python@v6
       with:
         python-version: ${{ matrix.python-version }}
 
@@ -93,6 +93,7 @@ jobs:
     - name: Install Carolina from wheel
       run: |
         source othervenv/bin/activate
+        pip install -U pip
         pyver_nodot=$(echo ${{ matrix.python-version }} | sed 's/\.//g')
         wheel_path=$(find carolina_dist -name "carolina*$pyver_nodot*whl")
         echo "Found Carolina wheel at $wheel_path"
@@ -103,12 +104,12 @@ jobs:
         source othervenv/bin/activate
         pip install pytest numpy
         cd carolina-main/tests
-        python -m pytest .
+        pytest . -sv
         cd -
 
     - name: Upload wheel as artifact
       if: always()
-      uses: actions/upload-artifact@v4
+      uses: actions/upload-artifact@v7
       with:
         name: ${{ matrix.os }} Python ${{ matrix.python-version }} wheel
         path: |

--- a/.github/workflows/bundle_with_dakota_macos.yml
+++ b/.github/workflows/bundle_with_dakota_macos.yml
@@ -29,14 +29,15 @@ jobs:
       fail-fast: true
       matrix:
         # https://github.com/actions/runner-images?tab=readme-ov-file#available-images
-        python-version: [ '3.11', '3.12', '3.13']
-        os: [ 'macos-14' ]
+        python-version: [ '3.11', '3.12', '3.13', '3.14' ]
+        os: [ 'macos-15' ]
 
     runs-on: ${{ matrix.os }}
     name: "Build 🛞 (${{ matrix.python-version }}, ${{ matrix.os }})"
     env:
       INSTALL_DIR: ${{ inputs.INSTALL_DIR }}
       SYSTEM_VERSION_COMPAT: 0
+      MACOSX_DEPLOYMENT_TARGET: '15.0'
     steps:
 
     - name: Echo env and inputs
@@ -53,7 +54,7 @@ jobs:
 
     - name: Cache Homebrew Bundler RubyGems
       id: cache
-      uses: actions/cache@v4
+      uses: actions/cache@v5
       with:
         path: ${{ steps.set-up-homebrew.outputs.gems-path }}
         key: ${{ runner.os }}-rubygems-${{ steps.set-up-homebrew.outputs.gems-hash }}
@@ -64,7 +65,7 @@ jobs:
       run: brew install-bundler-gems
 
     - name: Set up Python ${{ matrix.python-version }}
-      uses: actions/setup-python@v5
+      uses: actions/setup-python@v6
       with:
         python-version: ${{ matrix.python-version }}
 
@@ -77,14 +78,14 @@ jobs:
         which python
         which pip
         python -c "import platform; mac_version=platform.mac_ver(); print('macOS version: '+mac_version[0])"
-    - uses: actions/cache@v4
+    - uses: actions/cache@v5
       id: cache-package
       with:
         key: ${{ matrix.os }}_python-${{ matrix.python-version }}_boost-${{ inputs.BOOST_VERSION }}_dakota-${{ inputs.DAKOTA_VERSION }}
         path: ${{ github.workspace }}/python-${{ matrix.python-version}}_boost-${{ inputs.BOOST_VERSION }}_dakota-${{ inputs.DAKOTA_VERSION }}.tar.gz
 
     - name: Checkout Carolina
-      uses: actions/checkout@v4
+      uses: actions/checkout@v6
       with:
         fetch-depth: 0
 
@@ -96,6 +97,7 @@ jobs:
       run: |
         python -m pip install -U pip
         python -m pip install numpy
+        python -m pip install "cmake<4"
 
     - name: Build needed boost libraries
       if: steps.cache-package.outputs.cache-hit != 'true'
@@ -162,15 +164,23 @@ jobs:
         patch -p1 < $CAROLINA_DIR/dakota_manylinux_install_files/workdirhelper_boost_filesystem.patch
         patch -p1 < $CAROLINA_DIR/dakota_manylinux_install_files/CMakeLists_includes.patch
 
+        # Fix JEGA keyed_registry.hpp.inl bug (upstream fix: dakota-packages commit b837a87)
+        JEGA_KEYED_REG=$(find . -path '*/JEGA/eddy/utilities/include/inline/keyed_registry.hpp.inl')
+        if [ -n "$JEGA_KEYED_REG" ]; then
+          echo "Applying JEGA keyed_registry fix to $JEGA_KEYED_REG ..."
+          sed -i '' 's/const KeyType& value/const KeyType\& key/' "$JEGA_KEYED_REG"
+          sed -i '' 's/this->find(this->key)/this->find(key)/' "$JEGA_KEYED_REG"
+        fi
+
         PYTHON_INCLUDE_DIR="$(python -c "from sysconfig import get_paths as gp; g=gp(); print(f\"{g['include']} \")")"
 
         mkdir -p build
         cd build
 
         cmake \
-          -DCMAKE_CXX_STANDARD=14 \
+          -DCMAKE_CXX_STANDARD=17 \
           -DBUILD_SHARED_LIBS=ON \
-          -DCMAKE_CXX_FLAGS="-I$PYTHON_INCLUDE_DIR" \
+          -DCMAKE_CXX_FLAGS="-I$PYTHON_INCLUDE_DIR -D_LIBCPP_ENABLE_CXX17_REMOVED_UNARY_BINARY_FUNCTION -D_LIBCPP_ENABLE_CXX17_REMOVED_BINDERS" \
           -DDAKOTA_PYTHON_DIRECT_INTERFACE=ON \
           -DDAKOTA_PYTHON_DIRECT_INTERFACE_NUMPY=ON \
           -DDAKOTA_DLL_API=OFF \
@@ -192,12 +202,12 @@ jobs:
 
     - name: Cache boost and dakota
       if: steps.cache-package.outputs.cache-hit != 'true'
-      uses: actions/cache/save@v4
+      uses: actions/cache/save@v5
       with:
         path: ${{ github.workspace }}/python-${{ matrix.python-version}}_boost-${{ inputs.BOOST_VERSION }}_dakota-${{ inputs.DAKOTA_VERSION }}.tar.gz
         key: ${{ matrix.os }}_python-${{ matrix.python-version }}_boost-${{ inputs.BOOST_VERSION }}_dakota-${{ inputs.DAKOTA_VERSION }}
 
-    - uses: actions/cache/restore@v4
+    - uses: actions/cache/restore@v5
       id: restore-cached-package
       with:
         key: ${{ matrix.os }}_python-${{ matrix.python-version }}_boost-${{ inputs.BOOST_VERSION }}_dakota-${{ inputs.DAKOTA_VERSION }}
@@ -219,7 +229,7 @@ jobs:
         echo "DYLD_LIBRARY_PATH=${DYLD_LIBRARY_PATH}:${HOME}/${INSTALL_DIR}/lib:$INSTALL_DIR/bin" >> ${GITHUB_ENV}
 
     - name: Checkout Carolina
-      uses: actions/checkout@v4
+      uses: actions/checkout@v6
       with:
         fetch-depth: 0
 
@@ -230,17 +240,10 @@ jobs:
         arch_ver=$(uname -m)
 
         # hint regarding architecture to avoid universal wheel generated
-        export _PYTHON_HOST_PLATFORM="macosx-11.0-$arch_ver"
+        export _PYTHON_HOST_PLATFORM="macosx-${MACOSX_DEPLOYMENT_TARGET}-$arch_ver"
         export ARCHFLAGS="-arch $arch_ver"
 
-        target_version=$(sw_vers --productVersion | cut -d . -f 1)
-
-        # the arm64 builds require deployment targets >= 13.0
-        if [ "$arch_ver" == "arm64" ]; then
-          export MACOSX_DEPLOYMENT_TARGET="${target_version}.0"
-        fi
-
-        pip install delocate==0.10.7
+        pip install delocate
         python -m pip install .
         PYTHON_VERSION_TRIM=$(echo ${{ matrix.python-version }} | tr -d ".")
         echo $PYTHON_VERSION_TRIM
@@ -293,7 +296,7 @@ jobs:
         DYLD_LIBRARY_PATH=${EXTENDED_DYLD_LIBRARY_PATH} delocate-wheel -w /tmp/carolina_dist -v $unfixed_wheel_path 
         echo "Output new carolina wheel to /tmp/carolina_dist"
         ls -lh /tmp/carolina_dist
-    - uses: actions/upload-artifact@v4
+    - uses: actions/upload-artifact@v7
       id: cache-carolina_wheel
       with:
         name: ${{ matrix.os }}_py-${{ matrix.python-version }}_wheel
@@ -309,17 +312,17 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: [ '3.11', '3.12', '3.13']
-        os: [ 'macos-14' ]
+        python-version: [ '3.11', '3.12', '3.13', '3.14' ]
+        os: [ 'macos-15' ]
 
     runs-on: ${{ matrix.os }}
     steps:
-    - uses: actions/checkout@v4
+    - uses: actions/checkout@v6
       with:
         fetch-depth: 0
 
     - name: Set up Python ${{ matrix.python-version }}
-      uses: actions/setup-python@v5
+      uses: actions/setup-python@v6
       with:
         python-version: ${{ matrix.python-version }}
         cache: "pip"
@@ -328,13 +331,14 @@ jobs:
           pyproject.toml
 
     - name: Download Carolina artifact
-      uses: actions/download-artifact@v4
+      uses: actions/download-artifact@v8
       with:
         path: /tmp/artifacts
 
     - name: Find the Carolina wheel
       run: |
         python --version
+        pip install -U pip
 
         pyver_nodot=$(echo ${{ matrix.python-version }} | sed 's/\.//g')
         arch_ver=$(uname -m)

--- a/dakota_manylinux_install_files/CMakeLists.txt.patch
+++ b/dakota_manylinux_install_files/CMakeLists.txt.patch
@@ -1,6 +1,6 @@
 --- a/CMakeLists.txt	2023-11-24 10:57:08
 +++ b/CMakeLists.txt	2023-11-24 15:13:47
-@@ -150,10 +150,19 @@
+@@ -150,10 +150,15 @@
  # Perl is required for examples, docs, and system tests
  find_package(Perl REQUIRED)
  
@@ -9,18 +9,14 @@
  # for components here at top-level:
 +include_directories(${PYTHON_INCLUDE_DIRS})
 +include_directories(/usr/include)
-+include_directories(/usr/lib64)
-+include_directories(/tmp/INSTALL_DIR/dakota-6.21.0-public-src-cli/packages/external/eigen3/include/)
 +find_package(Threads REQUIRED)
-+set (PYBIND11_PYTHON_VERSION "3.11")
 +
  include(DakotaFindPython)
  dakota_find_python()
-+add_link_options()
  
  # Conditionally find Java JDK needed for input spec, docs, Java API
  include(DakotaFindJava)
-@@ -194,6 +203,8 @@
+@@ -194,6 +199,8 @@
    endif()
  endif()
  

--- a/dakota_manylinux_install_files/build_deps_docker.sh
+++ b/dakota_manylinux_install_files/build_deps_docker.sh
@@ -1,10 +1,13 @@
 #!/bin/bash
-set -e
+set -euo pipefail
 
-if [ -z "$1" ]; then
+
+if [ -z "${1:-}" ]; then
   echo "Please provide a Python version as an argument (e.g., 3.10)"
   exit 1
 fi
+
+PYTHON_VERSION_ARG="$1"
 
 BOOST_VERSION="1.87.0"
 DAKOTA_VERSION="6.21.0"
@@ -15,167 +18,169 @@ INSTALL_DIR=/tmp/INSTALL_DIR
 WORKSPACE=/tmp
 # WORKSPACE=/github/workspace
 
-mkdir -p $INSTALL_DIR
-mkdir -p $WORKSPACE/trace
-touch $WORKSPACE/trace/boost_bootstrap.log
-touch $WORKSPACE/trace/boost_install.log
-touch $WORKSPACE/trace/dakota_bootstrap.log
-touch $WORKSPACE/trace/dakota_install.log
-touch $WORKSPACE/trace/env
+mkdir -p "$INSTALL_DIR"
+mkdir -p "$WORKSPACE/trace"
+touch "$WORKSPACE/trace/boost_bootstrap.log"
+touch "$WORKSPACE/trace/boost_install.log"
+touch "$WORKSPACE/trace/dakota_bootstrap.log"
+touch "$WORKSPACE/trace/dakota_install.log"
+touch "$WORKSPACE/trace/env"
 
 # VERY IMPORTANT: extract python dev headers,
 # more info: https://github.com/pypa/manylinux/pull/1250
 pushd /opt/_internal && tar -xJf static-libs-for-embedding-only.tar.xz && popd
 
-echo "pushd /opt/_internal && tar -xJf static-libs-for-embedding-only.tar.xz && popd" >> $WORKSPACE/trace/env
-echo "INSTALL_DIR=$INSTALL_DIR" >> $WORKSPACE/trace/env
+echo "pushd /opt/_internal && tar -xJf static-libs-for-embedding-only.tar.xz && popd" >> "$WORKSPACE/trace/env"
+echo "INSTALL_DIR=$INSTALL_DIR" >> "$WORKSPACE/trace/env"
 
 yum install lapack-devel -y
 yum install python3-devel.x86_64 -y
 yum install -y wget
 cd /tmp
 
-BOOST_VERSION_UNDERSCORES=$(echo $BOOST_VERSION | sed 's/\./_/g')
+BOOST_VERSION_UNDERSCORES=$(echo "$BOOST_VERSION" | sed 's/\./_/g')
 
-if [ ! -f boost_${BOOST_VERSION_UNDERSCORES}.tar.bz2 ]; then
-  echo "Downloading boost archive .."
-  wget --quiet https://sourceforge.net/projects/boost/files/boost/${BOOST_VERSION}/boost_${BOOST_VERSION_UNDERSCORES}.tar.bz2 --no-check-certificate > /dev/null
+if [ ! -f "boost_${BOOST_VERSION_UNDERSCORES}.tar.bz2" ]; then
+  echo "Downloading boost archive ..."
+  wget --quiet "https://sourceforge.net/projects/boost/files/boost/${BOOST_VERSION}/boost_${BOOST_VERSION_UNDERSCORES}.tar.bz2" --no-check-certificate > /dev/null
 else
   echo "Found cached boost archive!"
 fi
 
-python_exec=$(which python$1)
-$python_exec -m venv myvenv
+python_exec=$(which "python${PYTHON_VERSION_ARG}")
+"$python_exec" -m venv myvenv
 source ./myvenv/bin/activate
 pip install numpy
-pip install pybind11[global]
+pip install "pybind11[global]"
+pip install "cmake<4"
 
 PYTHON_DEV_HEADERS_DIR=$(rpm -ql python3-devel.x86_64 | grep '\.h$' | head -n 1 | xargs dirname)
-NUMPY_INCLUDE_PATH=$(find /tmp -type d -path "*site-packages/numpy/core/include")
 PYTHON_INCLUDE_PATH=$(python -c "from sysconfig import get_paths; print(get_paths()['include'])")
 python_root=$(python -c "import sys; print(sys.prefix)")
-python_version=$(python --version | sed -E 's/.*([0-9]+\.[0-9]+)\.([0-9]+).*/\1/')
+python_version=$(python -c "import sys; print(f'{sys.version_info.major}.{sys.version_info.minor}')")
 python_version_no_dots="$(echo "${python_version//\./}")"
 python_bin_include_lib="    using python : $python_version : $(python -c "from sysconfig import get_paths as gp; g=gp(); print(f\"$python_exec : {g['include']} : {g['stdlib']} ;\")")"
 
 echo "Found dev headers $PYTHON_DEV_HEADERS_DIR"
-echo "Found numpy include path $NUMPY_INCLUDE_PATH"
 echo "Found python include path $PYTHON_INCLUDE_PATH"
 echo "Found python root $python_root"
 
-tar xf boost_$BOOST_VERSION_UNDERSCORES.tar.bz2
-cd boost_$BOOST_VERSION_UNDERSCORES
+tar xf "boost_${BOOST_VERSION_UNDERSCORES}.tar.bz2"
+cd "boost_${BOOST_VERSION_UNDERSCORES}"
 
-echo "python_exec=$python_exec" >> $WORKSPACE/trace/env
-echo "PYTHON_DEV_HEADERS_DIR=$PYTHON_DEV_HEADERS_DIR" >> $WORKSPACE/trace/env
-echo "NUMPY_INCLUDE_PATH=$NUMPY_INCLUDE_PATH" >> $WORKSPACE/trace/env
-echo "PYTHON_INCLUDE_PATH=$PYTHON_INCLUDE_PATH" >> $WORKSPACE/trace/env
-echo "python_root=$python_root" >> $WORKSPACE/trace/env
-echo "python_version=$python_version" >> $WORKSPACE/trace/env
-echo "python_version_no_dots=$python_version_no_dots" >> $WORKSPACE/trace/env
-echo "python_bin_include_lib=$python_bin_include_lib" >> $WORKSPACE/trace/env
-echo "bootstrap_cmd=./bootstrap.sh --with-libraries=python,filesystem,program_options,regex,serialization,system --with-python=$(which python) --with-python-root=$python_root &> "$INSTALL_DIR/boost_bootstrap.log"" >> $WORKSPACE/trace/env
+echo "python_exec=$python_exec" >> "$WORKSPACE/trace/env"
+echo "PYTHON_DEV_HEADERS_DIR=$PYTHON_DEV_HEADERS_DIR" >> "$WORKSPACE/trace/env"
+echo "PYTHON_INCLUDE_PATH=$PYTHON_INCLUDE_PATH" >> "$WORKSPACE/trace/env"
+echo "python_root=$python_root" >> "$WORKSPACE/trace/env"
+echo "python_version=$python_version" >> "$WORKSPACE/trace/env"
+echo "python_version_no_dots=$python_version_no_dots" >> "$WORKSPACE/trace/env"
+echo "python_bin_include_lib=$python_bin_include_lib" >> "$WORKSPACE/trace/env"
 
-./bootstrap.sh --with-libraries=python,filesystem,program_options,regex,serialization,system --with-python=$(which python) --with-python-root="$python_root" &> "$INSTALL_DIR/boost_bootstrap.log"
+./bootstrap.sh --with-libraries=python,filesystem,program_options,regex,serialization,system --with-python="$(which python)" --with-python-root="$python_root" &> "$INSTALL_DIR/boost_bootstrap.log"
 sed -i -e "s|.*using python.*|$python_bin_include_lib|" project-config.jam
-echo "# sed -i -e \"s|.*using python.*|$python_bin_include_lib|\" project-config.jam" >> $WORKSPACE/trace/env
 
-./b2 install -j8 -a cxxflags="-std=c++17" --prefix="$INSTALL_DIR" &> $WORKSPACE/trace/boost_install.log
-echo "# ./b2 install -j8 -a cxxflags="-std=c++17" --prefix="$INSTALL_DIR" &> $WORKSPACE/trace/boost_install.log" >> $WORKSPACE/trace/env
+./b2 install -j8 -a cxxflags="-std=c++17" --prefix="$INSTALL_DIR" &> "$WORKSPACE/trace/boost_install.log"
 
 cd /tmp
 
-if [ ! -f dakota-$DAKOTA_VERSION-public-src-cli.tar.gz ]; then
-  echo "Downloading Dakota archive .."
-  wget https://github.com/snl-dakota/dakota/releases/download/v$DAKOTA_VERSION/dakota-$DAKOTA_VERSION-public-src-cli.tar.gz > /dev/null
+if [ ! -f "dakota-${DAKOTA_VERSION}-public-src-cli.tar.gz" ]; then
+  echo "Downloading Dakota archive ..."
+  wget "https://github.com/snl-dakota/dakota/releases/download/v${DAKOTA_VERSION}/dakota-${DAKOTA_VERSION}-public-src-cli.tar.gz" > /dev/null
 else
   echo "Found cached Dakota archive!"
 fi
 
-cd $INSTALL_DIR
+cd "$INSTALL_DIR"
 DAKOTA_INSTALL_DIR=/tmp/INSTALL_DIR/dakota
-mkdir -p $DAKOTA_INSTALL_DIR
-echo "DAKOTA_INSTALL_DIR=$DAKOTA_INSTALL_DIR" >> $WORKSPACE/trace/env
+mkdir -p "$DAKOTA_INSTALL_DIR"
+echo "DAKOTA_INSTALL_DIR=$DAKOTA_INSTALL_DIR" >> "$WORKSPACE/trace/env"
 
-tar xf /tmp/dakota-$DAKOTA_VERSION-public-src-cli.tar.gz
+tar xf "/tmp/dakota-${DAKOTA_VERSION}-public-src-cli.tar.gz"
 
 CAROLINA_DIR=$WORKSPACE
 
-cd dakota-$DAKOTA_VERSION-public-src-cli
+cd "dakota-${DAKOTA_VERSION}-public-src-cli"
 
-patch -p1 < $CAROLINA_DIR/dakota_manylinux_install_files/CMakeLists.txt.patch
-patch -p1 < $CAROLINA_DIR/dakota_manylinux_install_files/DakotaFindPython.cmake.patch
-patch -p1 < $CAROLINA_DIR/dakota_manylinux_install_files/workdirhelper_boost_filesystem.patch
-patch -p1 < $CAROLINA_DIR/dakota_manylinux_install_files/CMakeLists_includes.patch
+echo "Applying patches ..."
+patch -p1 < "$CAROLINA_DIR/dakota_manylinux_install_files/CMakeLists.txt.patch"
+patch -p1 < "$CAROLINA_DIR/dakota_manylinux_install_files/DakotaFindPython.cmake.patch"
+patch -p1 < "$CAROLINA_DIR/dakota_manylinux_install_files/workdirhelper_boost_filesystem.patch"
+patch -p1 < "$CAROLINA_DIR/dakota_manylinux_install_files/CMakeLists_includes.patch"
+
+# FIX: GCC 14 no longer transitively provides std::uint32_t etc. via <string>.
+# Trilinos's Teuchos_BigUIntDecl.hpp needs an explicit #include <cstdint>.
+echo "Applying GCC 14 compatibility fix (cstdint) ..."
+sed -i '/#ifndef TEUCHOS_BIG_UINT_DECL_HPP/a #include <cstdint>' \
+  packages/external/trilinos/packages/teuchos/core/src/Teuchos_BigUIntDecl.hpp
+
+# Fix JEGA keyed_registry.hpp.inl bug (upstream fix: dakota-packages commit b837a87)
+JEGA_KEYED_REG=$(find . -path '*/JEGA/eddy/utilities/include/inline/keyed_registry.hpp.inl')
+if [ -n "$JEGA_KEYED_REG" ]; then
+  echo "Applying JEGA keyed_registry fix to $JEGA_KEYED_REG ..."
+  sed -i 's/const KeyType& value/const KeyType\& key/' "$JEGA_KEYED_REG"
+  sed -i 's/this->find(this->key)/this->find(key)/' "$JEGA_KEYED_REG"
+fi
+
 mkdir build
 cd build
 
-export PATH=/tmp/INSTALL_DIR/bin:$PATH
+export PATH="/tmp/INSTALL_DIR/bin:$PATH"
 export PYTHON_INCLUDE_DIRS="$PYTHON_INCLUDE_PATH $PYTHON_DEV_HEADERS_DIR /tmp/INSTALL_DIR/lib"
-export PYTHON_EXECUTABLE=$(which python)
-export LD_LIBRARY_PATH="$INSTALL_DIR/lib:/usr/local/lib:$PYTHON_INCLUDE_PATH:$NUMPY_INCLUDE_PATH:$NUMPY_INCLUDE_PATH/numpy:$PYTHON_DEV_HEADERS_DIR:/tmp/INSTALL_DIR/lib:$LD_LIBRARY_PATH"
+export PYTHON_EXECUTABLE="$(which python)"
+export LD_LIBRARY_PATH="$INSTALL_DIR/lib:/usr/local/lib:$PYTHON_DEV_HEADERS_DIR:/tmp/INSTALL_DIR/lib${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}"
 
-echo "export PATH=$PATH" >> $WORKSPACE/trace/env
-echo "export LD_LIBRARY_PATH=$LD_LIBRARY_PATH" >> $WORKSPACE/trace/env
-echo "export PYTHON_INCLUDE_DIRS=$PYTHON_INCLUDE_DIRS" >> $WORKSPACE/trace/env
-echo "export PYTHON_EXECUTABLE=$PYTHON_EXECUTABLE" >> $WORKSPACE/trace/env
+echo "export PATH=$PATH" >> "$WORKSPACE/trace/env"
+echo "export LD_LIBRARY_PATH=$LD_LIBRARY_PATH" >> "$WORKSPACE/trace/env"
+echo "export PYTHON_INCLUDE_DIRS=$PYTHON_INCLUDE_DIRS" >> "$WORKSPACE/trace/env"
+echo "export PYTHON_EXECUTABLE=$PYTHON_EXECUTABLE" >> "$WORKSPACE/trace/env"
 
-export BOOST_PYTHON="boost_python$python_version_no_dots"
-export BOOST_ROOT=$INSTALL_DIR
+export BOOST_PYTHON="boost_python${python_version_no_dots}"
+export BOOST_ROOT="$INSTALL_DIR"
 export PATH="$PATH:$INSTALL_DIR/bin"
 
 # More stable approach: Go via python
-numpy_lib_dir=$(find /tmp/myvenv/ -name numpy.libs)
-export LD_LIBRARY_PATH="/usr/lib:/usr/lib64:$INSTALL_DIR/lib:$INSTALL_DIR/bin:$numpy_lib_dir:$NUMPY_INCLUDE_PATH"
-export CMAKE_LIBRARY_PATH=$(echo $LD_LIBRARY_PATH | sed 's/::/:/g' | sed 's/:/;/g')
-export PYTHON_LIBRARIES="/usr/lib64/"
-export PYTHON_INCLUDE_DIR="/opt/_internal/cpython-3.7.17/include/python3.7m"
-export CMAKE_LINK_OPTS="-Wl,--copy-dt-needed-entries,-l pthread"
+numpy_lib_dir=$(find /tmp/myvenv/ -name numpy.libs 2>/dev/null || echo "")
+export PYTHON_LIBRARIES=$(python -c "import sysconfig; print(sysconfig.get_config_var('LIBDIR'))")
+export LD_LIBRARY_PATH="/usr/lib:/usr/lib64:$INSTALL_DIR/lib:$INSTALL_DIR/bin${numpy_lib_dir:+:$numpy_lib_dir}${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}"
+export CMAKE_LIBRARY_PATH=$(echo "$LD_LIBRARY_PATH" | sed 's/::/:/g' | sed 's/:/;/g')
+export PYTHON_INCLUDE_DIR="$PYTHON_DEV_HEADERS_DIR"
 
-export PYTHON_INCLUDE_DIR=$PYTHON_DEV_HEADERS_DIR
-
-echo "export BOOST_PYTHON=$BOOST_PYTHON" >> $WORKSPACE/trace/env
-echo "export BOOST_ROOT=$BOOST_ROOT" >> $WORKSPACE/trace/env
-echo "export PATH=$PATH" >> $WORKSPACE/trace/env
-echo "export LD_LIBRARY_PATH=$LD_LIBRARY_PATH" >> $WORKSPACE/trace/env
-echo "export CMAKE_LIBRARY_PATH=\"$CMAKE_LIBRARY_PATH\"" >> $WORKSPACE/trace/env
-echo "export PYTHON_LIBRARIES=\"$PYTHON_LIBRARIES\"" >> $WORKSPACE/trace/env
-echo "export PYTHON_INCLUDE_DIR=\"$PYTHON_INCLUDE_DIR\"" >> $WORKSPACE/trace/env
-echo "export CMAKE_LINK_OPTS=\"$CMAKE_LINK_OPTS\"" >> $WORKSPACE/trace/env
-
-cmake_command="""
-cmake \
-      -DCMAKE_CXX_STANDARD=14 \
-      -DBUILD_SHARED_LIBS=ON \
-      -DCMAKE_CXX_FLAGS=\"-I$PYTHON_INCLUDE_DIR\" \
-      -DDAKOTA_PYTHON_DIRECT_INTERFACE=ON \
-      -DDAKOTA_PYTHON_DIRECT_INTERFACE_NUMPY=ON \
-      -DDAKOTA_DLL_API=OFF \
-      -DHAVE_X_GRAPHICS=OFF \
-      -DDAKOTA_ENABLE_TESTS=OFF \
-      -DDAKOTA_ENABLE_TPL_TESTS=OFF \
-      -DCMAKE_BUILD_TYPE="Release" \
-      -DDAKOTA_NO_FIND_TRILINOS:BOOL=TRUE \
-      -DCMAKE_INSTALL_PREFIX="$INSTALL_DIR" \
-      -DPYTHON_LIBRARIES=$PYTHON_LIBRARIES \
-      -DCMAKE_LINK_OPTIONS=\"$CMAKE_LINK_OPTS\" \
-      .. &> "$INSTALL_DIR/dakota_bootstrap.log"
-
-"""
-echo "# $cmake_command" >> $WORKSPACE/trace/env
+echo "export BOOST_PYTHON=$BOOST_PYTHON" >> "$WORKSPACE/trace/env"
+echo "export BOOST_ROOT=$BOOST_ROOT" >> "$WORKSPACE/trace/env"
+echo "export PATH=$PATH" >> "$WORKSPACE/trace/env"
+echo "export LD_LIBRARY_PATH=$LD_LIBRARY_PATH" >> "$WORKSPACE/trace/env"
+echo "export CMAKE_LIBRARY_PATH=\"$CMAKE_LIBRARY_PATH\"" >> "$WORKSPACE/trace/env"
+echo "export PYTHON_LIBRARIES=\"$PYTHON_LIBRARIES\"" >> "$WORKSPACE/trace/env"
+echo "export PYTHON_INCLUDE_DIR=\"$PYTHON_INCLUDE_DIR\"" >> "$WORKSPACE/trace/env"
 
 echo "Boostrapping Dakota ..."
-$($cmake_command &> $WORKSPACE/trace/dakota_bootstrap.log)
+cmake \
+  -DCMAKE_CXX_STANDARD=17 \
+  -DBUILD_SHARED_LIBS=ON \
+  -DCMAKE_CXX_FLAGS="-I$PYTHON_INCLUDE_DIR" \
+  -DDAKOTA_PYTHON_DIRECT_INTERFACE=ON \
+  -DDAKOTA_PYTHON_DIRECT_INTERFACE_NUMPY=ON \
+  -DDAKOTA_DLL_API=OFF \
+  -DHAVE_X_GRAPHICS=OFF \
+  -DDAKOTA_ENABLE_TESTS=OFF \
+  -DDAKOTA_ENABLE_TPL_TESTS=OFF \
+  -DCMAKE_BUILD_TYPE="Release" \
+  -DDAKOTA_NO_FIND_TRILINOS:BOOL=TRUE \
+  -DCMAKE_INSTALL_PREFIX="$INSTALL_DIR" \
+  -DPYTHON_LIBRARIES="$PYTHON_LIBRARIES" \
+  -DCMAKE_EXE_LINKER_FLAGS="-L${PYTHON_LIBRARIES} -lpython${PYTHON_VERSION_ARG} -lpthread" \
+  -DTHREADS_PREFER_PTHREAD_FLAG=ON \
+  .. &> "$WORKSPACE/trace/dakota_bootstrap.log"
 
-echo "# make --debug=b -j8 install" >> $WORKSPACE/trace/env
 echo "Building Dakota ..."
-make --debug=b -j8 install &> $WORKSPACE/trace/dakota_install.log
+make --debug=b -j8 install &> "$WORKSPACE/trace/dakota_install.log"
 
 
-DEPS_BUILD=$WORKSPACE/deps_build
+DEPS_BUILD="$WORKSPACE/deps_build"
 
-mkdir -p $DEPS_BUILD
+mkdir -p "$DEPS_BUILD"
 
-cp -r $INSTALL_DIR/bin $DEPS_BUILD
-cp -r $INSTALL_DIR/lib $DEPS_BUILD
-cp -r $INSTALL_DIR/include $DEPS_BUILD
+cp -r "$INSTALL_DIR/bin" "$DEPS_BUILD"
+cp -r "$INSTALL_DIR/lib" "$DEPS_BUILD"
+cp -r "$INSTALL_DIR/include" "$DEPS_BUILD"
+

--- a/dakota_manylinux_install_files/build_deps_gha.sh
+++ b/dakota_manylinux_install_files/build_deps_gha.sh
@@ -1,16 +1,19 @@
 #!/bin/bash
-set -e
+set -euo pipefail
 
-if [ -z "$1" ]; then
+
+if [ -z "${1:-}" ]; then
   echo "Please provide a Python version as an argument (e.g., 3.10)"
   exit 1
 fi
 
+PYTHON_VERSION_ARG="$1"
+
 cd /tmp
 INSTALL_DIR=/tmp/INSTALL_DIR
 
-mkdir -p $INSTALL_DIR
-mkdir /github/workspace/trace
+mkdir -p "$INSTALL_DIR"
+mkdir -p /github/workspace/trace
 touch /github/workspace/trace/boost_bootstrap.log
 touch /github/workspace/trace/boost_install.log
 touch /github/workspace/trace/dakota_bootstrap.log
@@ -27,88 +30,100 @@ echo "INSTALL_DIR=$INSTALL_DIR" >> /github/workspace/trace/env
 yum install lapack-devel wget -y
 cd /tmp
 
-BOOST_VERSION_UNDERSCORES=$(echo $BOOST_VERSION | sed 's/\./_/g')
-wget --quiet https://sourceforge.net/projects/boost/files/boost/${BOOST_VERSION}/boost_${BOOST_VERSION_UNDERSCORES}.tar.bz2 --no-check-certificate > /dev/null
+if [ -z "${BOOST_VERSION:-}" ]; then
+  echo "ERROR: BOOST_VERSION environment variable is not set"
+  exit 1
+fi
+if [ -z "${DAKOTA_VERSION:-}" ]; then
+  echo "ERROR: DAKOTA_VERSION environment variable is not set"
+  exit 1
+fi
+
+BOOST_VERSION_UNDERSCORES=$(echo "$BOOST_VERSION" | sed 's/\./_/g')
+wget --quiet "https://sourceforge.net/projects/boost/files/boost/${BOOST_VERSION}/boost_${BOOST_VERSION_UNDERSCORES}.tar.bz2" --no-check-certificate > /dev/null
 
 # Install uv
 curl -LsSf https://astral.sh/uv/install.sh | sh
-uv python install $1
+uv python install "$PYTHON_VERSION_ARG"
 
-uv venv -p $1 myvenv
+uv venv -p "$PYTHON_VERSION_ARG" myvenv
 source ./myvenv/bin/activate
 uv pip install numpy
-uv pip install pybind11[global]
+uv pip install "pybind11[global]"
 uv pip install "cmake<4"
 
-NUMPY_INCLUDE_PATH=$(python3 -c "import numpy; print(numpy.get_include())")
 PYTHON_INCLUDE_PATH=$(python -c "from sysconfig import get_paths; print(get_paths()['include'])")
 python_root=$(python -c "import sys; print(sys.prefix)")
-python_version=$(python --version | sed -E 's/.*([0-9]+\.[0-9]+)\.([0-9]+).*/\1/')
+python_version=$(python -c "import sys; print(f'{sys.version_info.major}.{sys.version_info.minor}')")
 python_version_no_dots="$(echo "${python_version//\./}")"
+python_exec=$(which python)
 python_bin_include_lib="    using python : $python_version : $(python -c "from sysconfig import get_paths as gp; g=gp(); print(f\"$python_exec : {g['include']} : {g['stdlib']} ;\")")"
-PYTHON_INCLUDE_DIR="$(python -c "from sysconfig import get_paths as gp; g=gp(); print(f\"{g['include']} \")")"
+PYTHON_INCLUDE_DIR="$(python -c "from sysconfig import get_paths as gp; g=gp(); print(g['include'])")"
 
-echo "Found dev headers $PYTHON_DEV_HEADERS_DIR"
-echo "Found numpy include path $NUMPY_INCLUDE_PATH"
 echo "Found python include path $PYTHON_INCLUDE_PATH"
 echo "Found python root $python_root"
 echo "Found PYTHON_INCLUDE_DIR $PYTHON_INCLUDE_DIR"
 
-tar xf boost_$BOOST_VERSION_UNDERSCORES.tar.bz2
-cd boost_$BOOST_VERSION_UNDERSCORES
+tar xf "boost_${BOOST_VERSION_UNDERSCORES}.tar.bz2"
+cd "boost_${BOOST_VERSION_UNDERSCORES}"
 
-echo "NUMPY_INCLUDE_PATH=$NUMPY_INCLUDE_PATH" >> /github/workspace/trace/env
 echo "PYTHON_INCLUDE_PATH=$PYTHON_INCLUDE_PATH" >> /github/workspace/trace/env
 echo "python_root=$python_root" >> /github/workspace/trace/env
 echo "python_version=$python_version" >> /github/workspace/trace/env
 echo "python_version_no_dots=$python_version_no_dots" >> /github/workspace/trace/env
 echo "python_bin_include_lib=$python_bin_include_lib" >> /github/workspace/trace/env
-echo "bootstrap_cmd=./bootstrap.sh --with-libraries=python,filesystem,program_options,regex,serialization,system --with-python=$(which python) --with-python-root=$python_root &> "$INSTALL_DIR/boost_bootstrap.log"" >> /github/workspace/trace/env
 
-./bootstrap.sh --with-libraries=python,filesystem,program_options,regex,serialization,system --with-python=$(which python) --with-python-root="$python_root" &> "$INSTALL_DIR/boost_bootstrap.log"
+./bootstrap.sh --with-libraries=python,filesystem,program_options,regex,serialization,system --with-python="$python_exec" --with-python-root="$python_root" 2>&1 | tee "$INSTALL_DIR/boost_bootstrap.log"
 sed -i -e "s|.*using python.*|$python_bin_include_lib|" project-config.jam
-echo "# sed -i -e \"s|.*using python.*|$python_bin_include_lib|\" project-config.jam" >> /github/workspace/trace/env
 
-./b2 install -j8 -a cxxflags="-std=c++17" --prefix="$INSTALL_DIR" &> /github/workspace/trace/boost_install.log
-echo "# ./b2 install -j8 -a cxxflags="-std=c++17" --prefix="$INSTALL_DIR" &> /github/workspace/trace/boost_install.log" >> /github/workspace/trace/env
-
-cd $INSTALL_DIR
+./b2 install -j8 -a cxxflags="-std=c++17" --prefix="$INSTALL_DIR" 2>&1 | tee /github/workspace/trace/boost_install.log
+cd "$INSTALL_DIR"
 DAKOTA_INSTALL_DIR=/tmp/INSTALL_DIR/dakota
-mkdir -p $DAKOTA_INSTALL_DIR
+mkdir -p "$DAKOTA_INSTALL_DIR"
 echo "DAKOTA_INSTALL_DIR=$DAKOTA_INSTALL_DIR" >> /github/workspace/trace/env
 
-wget https://github.com/snl-dakota/dakota/releases/download/v$DAKOTA_VERSION/dakota-$DAKOTA_VERSION-public-src-cli.tar.gz > /dev/null
-tar xf dakota-$DAKOTA_VERSION-public-src-cli.tar.gz
+wget "https://github.com/snl-dakota/dakota/releases/download/v${DAKOTA_VERSION}/dakota-${DAKOTA_VERSION}-public-src-cli.tar.gz" > /dev/null
+tar xf "dakota-${DAKOTA_VERSION}-public-src-cli.tar.gz"
 
 CAROLINA_DIR=/github/workspace
-cd dakota-$DAKOTA_VERSION-public-src-cli
+cd "dakota-${DAKOTA_VERSION}-public-src-cli"
 
-patch -p1 < $CAROLINA_DIR/dakota_manylinux_install_files/CMakeLists.txt.patch
-patch -p1 < $CAROLINA_DIR/dakota_manylinux_install_files/DakotaFindPython.cmake.patch
-patch -p1 < $CAROLINA_DIR/dakota_manylinux_install_files/workdirhelper_boost_filesystem.patch
-patch -p1 < $CAROLINA_DIR/dakota_manylinux_install_files/CMakeLists_includes.patch
+echo "Applying patches ..."
+patch -p1 < "$CAROLINA_DIR/dakota_manylinux_install_files/CMakeLists.txt.patch"
+patch -p1 < "$CAROLINA_DIR/dakota_manylinux_install_files/DakotaFindPython.cmake.patch"
+patch -p1 < "$CAROLINA_DIR/dakota_manylinux_install_files/workdirhelper_boost_filesystem.patch"
+patch -p1 < "$CAROLINA_DIR/dakota_manylinux_install_files/CMakeLists_includes.patch"
+
+# FIX: GCC 14 no longer transitively provides std::uint32_t etc. via <string>.
+# Trilinos's Teuchos_BigUIntDecl.hpp needs an explicit #include <cstdint>.
+echo "Applying GCC 14 compatibility fix (cstdint) ..."
+sed -i '/#ifndef TEUCHOS_BIG_UINT_DECL_HPP/a #include <cstdint>' \
+  packages/external/trilinos/packages/teuchos/core/src/Teuchos_BigUIntDecl.hpp
+
+# Fix JEGA keyed_registry.hpp.inl bug (upstream fix: dakota-packages commit b837a87)
+JEGA_KEYED_REG=$(find . -path '*/JEGA/eddy/utilities/include/inline/keyed_registry.hpp.inl')
+if [ -n "$JEGA_KEYED_REG" ]; then
+  echo "Applying JEGA keyed_registry fix to $JEGA_KEYED_REG ..."
+  sed -i 's/const KeyType& value/const KeyType\& key/' "$JEGA_KEYED_REG"
+  sed -i 's/this->find(this->key)/this->find(key)/' "$JEGA_KEYED_REG"
+fi
+
 mkdir build
 cd build
 
-export PATH=/tmp/INSTALL_DIR/bin:$PATH
+export PATH="/tmp/INSTALL_DIR/bin:$PATH"
 export PYTHON_INCLUDE_DIRS="$PYTHON_INCLUDE_PATH /tmp/INSTALL_DIR/lib"
-export PYTHON_EXECUTABLE=$(which python)
+export PYTHON_EXECUTABLE="$python_exec"
 
-echo "export PATH=$PATH" >> /github/workspace/trace/env
-echo "export PYTHON_INCLUDE_DIRS=$PYTHON_INCLUDE_DIRS" >> /github/workspace/trace/env
-echo "export PYTHON_EXECUTABLE=$PYTHON_EXECUTABLE" >> /github/workspace/trace/env
-
-export BOOST_PYTHON="boost_python$python_version_no_dots"
-export BOOST_ROOT=$INSTALL_DIR
+export BOOST_PYTHON="boost_python${python_version_no_dots}"
+export BOOST_ROOT="$INSTALL_DIR"
 export PATH="$PATH:$INSTALL_DIR/bin"
 
 # More stable approach: Go via python
-numpy_lib_dir=$(find /tmp/myvenv/ -name numpy.libs)
-export PYTHON_LIBRARIES=$(python3 -c "import sysconfig; print(sysconfig.get_config_var('LIBDIR'))")
-export LD_LIBRARY_PATH="$PYTHON_LIBRARIES:/usr/lib:/usr/lib64:$INSTALL_DIR/lib:$INSTALL_DIR/bin:/lib64:$numpy_lib_dir:$NUMPY_INCLUDE_PATH:$LD_LIBRARY_PATH"
-export CMAKE_LIBRARY_PATH=$(echo $LD_LIBRARY_PATH | sed 's/::/:/g' | sed 's/:/;/g')
-export CMAKE_LINK_OPTS="-Wl,--copy-dt-needed-entries, -lpthread"
-
+numpy_lib_dir=$(find /tmp/myvenv/ -name numpy.libs 2>/dev/null || echo "")
+export PYTHON_LIBRARIES=$(python -c "import sysconfig; print(sysconfig.get_config_var('LIBDIR'))")
+export LD_LIBRARY_PATH="$PYTHON_LIBRARIES:/usr/lib:/usr/lib64:$INSTALL_DIR/lib:$INSTALL_DIR/bin:/lib64${numpy_lib_dir:+:$numpy_lib_dir}${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}"
+export CMAKE_LIBRARY_PATH=$(echo "$LD_LIBRARY_PATH" | sed 's/::/:/g' | sed 's/:/;/g')
 
 echo "export BOOST_PYTHON=$BOOST_PYTHON" >> /github/workspace/trace/env
 echo "export BOOST_ROOT=$BOOST_ROOT" >> /github/workspace/trace/env
@@ -117,11 +132,11 @@ echo "export LD_LIBRARY_PATH=$LD_LIBRARY_PATH" >> /github/workspace/trace/env
 echo "export CMAKE_LIBRARY_PATH=\"$CMAKE_LIBRARY_PATH\"" >> /github/workspace/trace/env
 echo "export PYTHON_LIBRARIES=\"$PYTHON_LIBRARIES\"" >> /github/workspace/trace/env
 echo "export PYTHON_INCLUDE_DIR=\"$PYTHON_INCLUDE_DIR\"" >> /github/workspace/trace/env
-echo "export CMAKE_LINK_OPTS=\"$CMAKE_LINK_OPTS\"" >> /github/workspace/trace/env
+echo "export PYTHON_EXECUTABLE=\"$PYTHON_EXECUTABLE\"" >> /github/workspace/trace/env
 
 echo "Bootstrapping Dakota ..."
 cmake \
-  -DCMAKE_CXX_STANDARD=14 \
+  -DCMAKE_CXX_STANDARD=17 \
   -DBUILD_SHARED_LIBS=ON \
   -DCMAKE_CXX_FLAGS="-I$PYTHON_INCLUDE_DIR" \
   -DDAKOTA_PYTHON_DIRECT_INTERFACE=ON \
@@ -133,21 +148,20 @@ cmake \
   -DCMAKE_BUILD_TYPE="Release" \
   -DDAKOTA_NO_FIND_TRILINOS:BOOL=TRUE \
   -DCMAKE_INSTALL_PREFIX="$INSTALL_DIR" \
-  -DPYTHON_LIBRARIES=$PYTHON_LIBRARIES \
-  -DCMAKE_LINK_OPTIONS="$CMAKE_LINK_OPTS" \
-  -DCMAKE_EXE_LINKER_FLAGS="-L${PYTHON_LIBRARIES} -lpython$1" \
+  -DPYTHON_LIBRARIES="$PYTHON_LIBRARIES" \
+  -DCMAKE_EXE_LINKER_FLAGS="-L${PYTHON_LIBRARIES} -lpython${PYTHON_VERSION_ARG} -lpthread" \
   -DTHREADS_PREFER_PTHREAD_FLAG=ON \
-  .. &> /github/workspace/trace/dakota_bootstrap.log
+  .. 2>&1 | tee /github/workspace/trace/dakota_bootstrap.log
 
-echo "# make --debug=b -j8 install" >> /github/workspace/trace/env
 echo "Building Dakota ..."
-make --debug=b -j8 install &> /github/workspace/trace/dakota_install.log
-
+make -j8 install 2>&1 | tee /github/workspace/trace/dakota_install.log
 
 DEPS_BUILD=/github/workspace/deps_build
 
-mkdir -p $DEPS_BUILD
+mkdir -p "$DEPS_BUILD"
 
-cp -r $INSTALL_DIR/bin $DEPS_BUILD
-cp -r $INSTALL_DIR/lib $DEPS_BUILD
-cp -r $INSTALL_DIR/include $DEPS_BUILD
+cp -r "$INSTALL_DIR/bin" "$DEPS_BUILD"
+cp -r "$INSTALL_DIR/lib" "$DEPS_BUILD"
+cp -r "$INSTALL_DIR/include" "$DEPS_BUILD"
+
+

--- a/dakota_manylinux_install_files/build_wheels_gha.sh
+++ b/dakota_manylinux_install_files/build_wheels_gha.sh
@@ -1,37 +1,41 @@
 #!/bin/bash
-set -e
+set -eu
 
-if [ -z "$1" ]; then
+
+if [ -z "${1:-}" ]; then
   echo "Please provide a Python version as an argument (e.g., 3.10)"
   exit 1
 fi
+
+PYTHON_VERSION_ARG="$1"
 
 CAROLINA_DIR=$(pwd)
 INSTALL_DIR=/github/workspace/deps_build
 
 cd /tmp
 
-python_exec=$(which python$1)
-$python_exec -m venv myvenv
+python_exec=$(which "python${PYTHON_VERSION_ARG}")
+"$python_exec" -m venv myvenv
 source ./myvenv/bin/activate
 
+pip install -U pip
+pip install auditwheel
 pip install pytest numpy
-NUMPY_INCLUDE_PATH=$(find /tmp -type d -path "*site-packages/numpy/core/include")
-numpy_lib_dir=$(find /tmp/myvenv/ -name numpy.libs)
+numpy_lib_dir=$(find /tmp/myvenv/ -name numpy.libs 2>/dev/null || echo "")
 yum install lapack-devel -y
 
 export PATH="$PATH:$INSTALL_DIR/bin"
-export LD_LIBRARY_PATH="/usr/lib:/usr/lib64:$INSTALL_DIR/lib:$INSTALL_DIR/bin:$numpy_lib_dir:$NUMPY_INCLUDE_PATH"
+export LD_LIBRARY_PATH="/usr/lib:/usr/lib64:$INSTALL_DIR/lib:$INSTALL_DIR/bin${numpy_lib_dir:+:$numpy_lib_dir}${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}"
 
-cd $CAROLINA_DIR
+cd "$CAROLINA_DIR"
 
 pip install .
 pip list | grep carolina
 
 pytest tests
 
-mkdir /tmp/wheels
-mkdir /github/workspace/carolina_dist
+mkdir -p /tmp/wheels
+mkdir -p /github/workspace/carolina_dist
 
 pip wheel . -w wheelhouse
 auditwheel repair wheelhouse/* -w /tmp/wheels

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -26,6 +26,7 @@ classifiers = ['Intended Audience :: Science/Research',
                'Programming Language :: Python :: 3.11',
                'Programming Language :: Python :: 3.12',
                'Programming Language :: Python :: 3.13',
+               'Programming Language :: Python :: 3.14',
                ]
 
 [tool.setuptools_scm]

--- a/script/build_boost_dakota.sh
+++ b/script/build_boost_dakota.sh
@@ -1,11 +1,12 @@
 #! /bin/bash
 # exit early if encounter failure
-set -e
+set -eu
+
 
 BOOST_VERSION="1.87.0"
 DAKOTA_VERSION="6.21.0"
 
-BOOST_VER_NODOTS=$(echo $BOOST_VERSION | sed 's/\./_/g')
+BOOST_VER_NODOTS=$(echo "$BOOST_VERSION" | sed 's/\./_/g')
 
 INSTALL_DIR="$(pwd)/everdeps"
 CACHE_DIR="$(pwd)/download_cache"
@@ -23,7 +24,7 @@ else
   exit
 fi
 
-VENV_BASE=$1
+VENV_BASE=${1:-}
 if [ -z "$VENV_BASE" ]; then
   echo "The virtual environment must be explicitly declared;"
   echo "Please supply the path of the virtual environment as the first argument"
@@ -50,9 +51,9 @@ write_to_setenv() {
 
 write_to_setenv INSTALL_DIR "$INSTALL_DIR"
 
-python_version=$(python --version | sed -E 's/.*([0-9]+\.[0-9]+)\.([0-9]+).*/\1/')
+python_version=$(python -c "import sys; print(f'{sys.version_info.major}.{sys.version_info.minor}')")
 python_bin_include_lib="    using python : $python_version : $(python -c "from sysconfig import get_paths as gp; g=gp(); print(f\"$(which python) : {g['include']} : {g['stdlib']} ;\")")"
-PYTHON_INCLUDE_DIR="$(python -c "from sysconfig import get_paths as gp; g=gp(); print(f\"{g['include']} \")")"
+PYTHON_INCLUDE_DIR="$(python -c "from sysconfig import get_paths as gp; g=gp(); print(g['include'])")"
 echo "Detected python version $python_version"
 
 write_to_setenv python_version "$python_version"
@@ -63,16 +64,16 @@ touch "$TRACE_DIR/boost_bootstrap.log"
 touch "$TRACE_DIR/boost_install.log"
 
 echo "Downloading and extracting Boost .."
-boost_file=boost_$BOOST_VER_NODOTS.tar.bz2
+boost_file="boost_${BOOST_VER_NODOTS}.tar.bz2"
 
 if [ ! -f "$CACHE_DIR/$boost_file" ]; then
-  wget --quiet https://sourceforge.net/projects/boost/files/boost/$BOOST_VERSION/boost_${$BOOST_VER_NODOTS}.tar.bz2 -P "$CACHE_DIR"
+  wget --quiet "https://sourceforge.net/projects/boost/files/boost/${BOOST_VERSION}/boost_${BOOST_VER_NODOTS}.tar.bz2" -P "$CACHE_DIR"
 else
   echo "Found $boost_file in download cache .."
 fi
 
 tar xf "$CACHE_DIR"/$boost_file
-cd boost_$BOOST_VER_NODOTS
+cd "boost_${BOOST_VER_NODOTS}"
 
 echo "Bootstrapping Boost .."
 ./bootstrap.sh --with-libraries=python,filesystem,program_options,regex,serialization,system --with-python-version="$python_version" &> "$TRACE_DIR/boost_bootstrap.log"
@@ -98,33 +99,41 @@ touch "$TRACE_DIR/dakota_install.log"
 # If it is installed under arm64 arch, this will fail,
 # solution is to uninstall and reinstall when in intel arch
 pip install numpy
-pip install cmake
+pip install "cmake<4"
 
 echo "Download and extract Dakota .."
 
-dakota_file=dakota-$DAKOTA_VERSION-public-src-cli.tar.gz
+dakota_file="dakota-${DAKOTA_VERSION}-public-src-cli.tar.gz"
 
 if [ ! -f "$CACHE_DIR"/$dakota_file ]; then
-  wget https://github.com/snl-dakota/dakota/releases/download/v$DAKOTA_VERSION/$dakota_file -P "$CACHE_DIR"
+  wget "https://github.com/snl-dakota/dakota/releases/download/v${DAKOTA_VERSION}/${dakota_file}" -P "$CACHE_DIR"
 else
   echo "Found $dakota_file in download cache .."
 fi
 
 tar xf "$CACHE_DIR"/$dakota_file
 
-cd dakota-$DAKOTA_VERSION-public-src-cli
+cd "dakota-${DAKOTA_VERSION}-public-src-cli"
 
 patch -p1 < ../../../dakota_manylinux_install_files/workdirhelper_boost_filesystem.patch
 patch -p1 < ../../../dakota_manylinux_install_files/CMakeLists_includes.patch
+
+# Fix JEGA keyed_registry.hpp.inl bug (upstream fix: dakota-packages commit b837a87)
+JEGA_KEYED_REG=$(find . -path '*/JEGA/eddy/utilities/include/inline/keyed_registry.hpp.inl')
+if [ -n "$JEGA_KEYED_REG" ]; then
+  echo "Applying JEGA keyed_registry fix to $JEGA_KEYED_REG ..."
+  sed -i '' 's/const KeyType& value/const KeyType\& key/' "$JEGA_KEYED_REG"
+  sed -i '' 's/this->find(this->key)/this->find(key)/' "$JEGA_KEYED_REG"
+fi
 
 mkdir -p build
 cd build
 
 echo "Building Dakota with cmake, logging to $TRACE_DIR/dakota_build.log"
 cmake \
-      -DCMAKE_CXX_STANDARD=14 \
+      -DCMAKE_CXX_STANDARD=17 \
       -DBUILD_SHARED_LIBS=ON \
-      -DCMAKE_CXX_FLAGS="-I$PYTHON_INCLUDE_DIR" \
+      -DCMAKE_CXX_FLAGS="-I$PYTHON_INCLUDE_DIR -D_LIBCPP_ENABLE_CXX17_REMOVED_UNARY_BINARY_FUNCTION -D_LIBCPP_ENABLE_CXX17_REMOVED_BINDERS" \
       -DDAKOTA_PYTHON_SURROGATES=ON \
       -DDAKOTA_PYTHON_DIRECT_INTERFACE=ON \
       -DDAKOTA_PYTHON_DIRECT_INTERFACE_NUMPY=ON \
@@ -171,25 +180,23 @@ write_to_setenv "export BOOST_PYTHON" "$BOOST_PYTHON"
 
 echo "Building Carolina .."
 pip install . &> "$TRACE_DIR/carolina_install.log"
-echo "Done"
 
 cd "$INSTALL_DIR"
 
 site_packages_dir=$(python -c "import site; print(site.getsitepackages()[0])")
-carolina_so_path=$(find "$site_packages_dir" -name "carolina.cpython-$python_version_no_dots-darwin.so")
+carolina_so_path=$(find "$site_packages_dir" -name "carolina.cpython-${python_version_no_dots}-darwin.so")
 fortran_dylib_path=$(find /usr -name "libgfortran.dylib" | head -n 1)
 
 echo "found site packages @ $site_packages_dir"
-echo "found carolina.cpython-$python_version_no_dots-darwin.so @ $carolina_so_path"
+echo "found carolina.cpython-${python_version_no_dots}-darwin.so @ $carolina_so_path"
 echo "found libgfortran.dylib @ $fortran_dylib_path"
 
 fortran_dylib_dir=$(dirname "$fortran_dylib_path")
 install_lib_dir="$INSTALL_DIR/lib"
 
-echo "Adding required rpaths to $carolina_so_path, like this:"
-
-echo "install_name_tool -add_rpath $fortran_dylib_dir $carolina_so_path"
-echo "install_name_tool -add_rpath $install_lib_dir $carolina_so_path"
+echo "Adding required rpaths to $carolina_so_path:"
+echo "  install_name_tool -add_rpath $fortran_dylib_dir $carolina_so_path"
+echo "  install_name_tool -add_rpath $install_lib_dir $carolina_so_path"
 
 install_name_tool -add_rpath "$fortran_dylib_dir" "$carolina_so_path"
 install_name_tool -add_rpath "$install_lib_dir" "$carolina_so_path"

--- a/setup.py
+++ b/setup.py
@@ -132,8 +132,11 @@ def get_carolina_extension():
 
     define_macros = get_define_macros(dakota_macros)
 
-    # macOS linker does not support this flag
-    extra_link_args = [] if "Darwin" in platform.system() else ['-Wl,-z origin']
+    if "Darwin" in platform.system():
+        # Reserve header space for install_name_tool rewrites by delocate
+        extra_link_args = ['-Wl,-headerpad_max_install_names']
+    else:
+        extra_link_args = ['-Wl,-z,origin']
 
     carolina = Extension(name='carolina',
                          sources=sources,

